### PR TITLE
Initial commit for multiprocessing firewall engine

### PR DESCRIPTION
**Pull request by VTabolin**

### CODE REVIEW

File: vmware_dvs/agent/dvs_neutron_agent.py [hunk 1]
```diff
@@ -65,7 +65,6 @@ def __init__(self, vsphere_hostname, vsphere_login, vsphere_password,
             'agent_type': 'DVS agent',
             'start_flag': True}
 
-        self.setup_rpc()
         report_interval = cfg.CONF.AGENT.report_interval
```
Comment (ekosareva) [inline vmware_dvs/agent/dvs_neutron_agent.py:68]: _report_state should be started after all rpc agents were initialized. So, please move L68-L72 under setup_rpc() function

Comment (VTabolin) [inline vmware_dvs/agent/dvs_neutron_agent.py:68]: Done

File: vmware_dvs/agent/firewalls/vcenter_firewall.py [hunk 2]
```diff
@@ -26,6 +31,56 @@
...
+    pq.port_updater_loop()
+    while True:
+        dvs, ports = pq.get_tasks(5)
+        if dvs and len(ports) > 0:
+            updater(dvs, ports)
+        else:
+            time.sleep(1)
+
+
+class port_queue(object):
```
Comment (ekosareva) [inline vmware_dvs/agent/firewalls/vcenter_firewall.py]: Regarding pep8 class names should normally use the CapWords convention. Please, change port_queue -> PortQueue

File: vmware_dvs/agent/firewalls/vcenter_firewall.py [hunk 3]
```diff
@@ -26,6 +31,58 @@
...
+        self.list_queues = list_queues
+        self.taskstore = {}
+        self.networking_map = dvs_util.create_network_map_from_config(
+            CONF.ML2_VMWARE)
+
+    # Todo: add roundrobin for active DVS. SlOPS
+    def get_tasks(self, number=0):
+        for dvs in self.taskstore:
+            tasks = self.taskstore.get(dvs, [])
+            if len(tasks) > 0:
```
Comment (ekosareva) [inline vmware_dvs/agent/firewalls/vcenter_firewall.py]: lines 54-56 is equal with:
for dvs, tasks in self.taskstore.iteritems():
    if tasks:

File: vmware_dvs/agent/firewalls/vcenter_firewall.py [hunk 4]
```diff
@@ -75,7 +138,29 @@ def _apply_sg_rules_for_port(self, ports):
...
+            for physnet, dvs_item in self.networking_map.iteritems():
+                if dvs_item == dvs:
+                    ports = [{'id': p['id'], 'security_group_rules':
+                            p['security_group_rules'], 'binding:vif_details':
+                            p['binding:vif_details']} for p in port_list]
+
+                    for port in ports:
+                        queue = self._getfreequeue()
+                        print queue
+                        queue.put({physnet: [port]})
```
Comment (ekosareva) [inline vmware_dvs/agent/firewalls/vcenter_firewall.py]: please, rewrite lines 143-150:

```
for port in port_list:
    queue = self._getfreequeue()
    queue.put({
        physnet: [{'id': port['id'],
                   'security_group_rules': port['security_group_rules'],
                   'binding:vif_details': port['binding:vif_details']}]
    })
```

File: vmware_dvs/agent/firewalls/vcenter_firewall.py [hunk 5]
```diff
@@ -75,7 +138,29 @@ def _apply_sg_rules_for_port(self, ports):
...
+        minimum = self.list_queues[0].qsize()
+        ret = self.list_queues[0]
+        for queue in self.list_queues:
+            if queue.qsize() < minimum:
+                minimum = queue.qsize()
+                ret = queue
+                if minimum == 0:
+                    return queue
+                ret = queue
+        return ret
```
Comment (ekosareva) [inline vmware_dvs/agent/firewalls/vcenter_firewall.py]:
```
shortest_queue = self.list_queues[0]
for queue in self.list_queues:
    queue_size = queue.qsize()
    if queue_size == 0:
        return queue
    if queue_size < shortest_queue.qsize():
        shortest_queue = queue
return shortest_queue         
```

File: vmware_dvs/utils/dvs_util.py [hunk 6]
```diff
@@ -325,6 +325,27 @@ def _get_ports_for_pg(self, pg_name):
 
     def _lookup_unbound_port(self, port_group):
         builder = SpecBuilder(self.connection.vim.client.factory)
+        criteria = builder.port_criteria(port_group_key=port_group.value,
+                                         connected=None)
```
Comment (ekosareva) [inline vmware_dvs/utils/dvs_util.py]: It's not obligatory pass `connected` param, by default connected=None

File: vmware_dvs/utils/dvs_util.py [hunk 7]
```diff
@@ -325,6 +325,27 @@ def _get_ports_for_pg(self, pg_name):
...
+                                         connected=True)
+        connected_port_keys = self.connection.invoke_api(
+            self.connection.vim,
+            'FetchDVPortKeys',
+            self._dvs, criteria=criteria)
+        free_keys = [x for x in all_port_keys if x not in connected_port_keys]
+        port_keys = [x for x in free_keys if x not in self._blocked_ports]
+        if len(port_keys) > 0:
+            self._blocked_ports.add(port_keys[0])
+            return self._get_port_info_by_portkey(port_keys[0])
```
Comment (ekosareva) [inline vmware_dvs/utils/dvs_util.py]: may be rewrite lines 340-345:

```
        for port_key in all_port_keys:
            if (port_key not in connected_port_keys and 
                    port_key not in self._blocked_ports):
                self._blocked_ports.add(port_key)
                return self._get_port_info_by_portkey(port_key)  
        raise exceptions.UnboundPortNotFound()
```

File: vmware_dvs/agent/firewalls/vcenter_firewall.py [hunk 8]
```diff
@@ -13,6 +13,11 @@
 #    License for the specific language governing permissions and limitations
 #    under the License.
 
+from multiprocessing import Process, Queue
+
```
Comment (ekosareva) [inline vmware_dvs/agent/firewalls/vcenter_firewall.py]: this empty line is not required

**Pull request merged**